### PR TITLE
Add a yosys/nextpnr-xilinx/prjxray toolchain option for the xilinx platform

### PR DESCRIPTION
**Pull request by LAK132**
This will be needed for the MEGA65 board files

### CODE REVIEW

File: torii/platform/vendor/xilinx.py [hunk 1]
```diff
@@ -430,6 +453,122 @@ def _part(self) -> str:
 		'''
 	]
 
+	# Yosys NextPNR prjxray templates
+
+	_yosys_nextpnr_part_map = {
+		'xc7a35ticsg324-1L':  'xc7a35tcsg324-1',  # Arty-A7 35t
+		'xc7a100ticsg324-1L': 'xc7a100tcsg324-1', # Arty-A7 100t
+
```
Comment (dragonmux) [inline torii/platform/vendor/xilinx.py:461]: This introduces a small whitespace error, please remove the extra empty line

File: torii/platform/vendor/xilinx.py [hunk 2]
```diff
@@ -430,6 +453,122 @@ def _part(self) -> str:
...
+		'xc7a': 'artix7',
+		'xc7z': 'zynq7',
+	}
+
+	_yosys_nextpnr_required_tools = [
+		'yosys',
+		'nextpnr-xilinx',
+		'fasm2frames',
+		'xc7frames2bit'
+	]
```
Comment (dragonmux) [inline torii/platform/vendor/xilinx.py:479]: Please add an extra line between the end of this  block and the start of the next

File: torii/platform/vendor/xilinx.py [hunk 3]
```diff
@@ -430,6 +453,122 @@ def _part(self) -> str:
...
+			{% for file in platform.iter_files(".il") -%}
+				read_ilang {{file}}
+			{% endfor %}
+			read_verilog {{name}}.v
+			{{get_override("script_after_read")|default("# (script_after_read placeholder)")}}
+			synth_xilinx {{get_override("synth_opts")|options}} -flatten -abc9 -nobram -arch xc7 -top {{name}};
+			{{get_override("script_after_synth")|default("# (script_after_synth placeholder)")}}
+			write_json {{name}}.json
+		'''
+	}
```
Comment (dragonmux) [inline torii/platform/vendor/xilinx.py:524]: This is missing a new line following the end of the dict

**Review by dragonmux — CHANGES_REQUESTED** [review]
This is looking very good - the only items we spotted in review are vertical whitespace issues. With those fixed, we'll merge this.

File: torii/platform/vendor/xilinx.py [hunk 4]
```diff
@@ -110,6 +110,29 @@ class XilinxPlatform(TemplatedPlatform):
...
+	.. rubric:: yosys_nextpnr toolchain
+
+	Required tools:
+		* ``yosys``
+		* ``nextpnr-xilinx``
+		* ``fasm2frames``
+		* ``xc7frames2bit``
+
+	The environment is populated by running the script specified in the environment variable
+	``AMARANTH_ENV_ysosys_nextpnr``, if present.
```
Comment (lethalbit) [inline torii/platform/vendor/xilinx.py:123]: This should be changed to `TORII_ENV_YOSYS_NEXTPNR`

File: torii/platform/vendor/xilinx.py [hunk 5]
```diff
@@ -110,6 +110,29 @@ class XilinxPlatform(TemplatedPlatform):
...
+		* ``nextpnr_db_dir``: overrides nextpnr-xilinx chip-db path.
+		* ``prjxray_db_dir``: overrides prjxray db path prefix.
+		* ``read_verilog_opts``: adds options for ``read_verilog`` Yosys command.
+		* ``script_after_read``: inserts commands after ``read_verilog`` in Yosys script.
+		* ``synth_opts``: adds options for ``synth_xilinx`` Yosys command.
+		* ``script_after_synth``: inserts commands after ``synth_xilinx`` in Yosys script.
+		* ``yosys_opts``: adds extra options for ``yosys``.
+		* ``nextpnr_opts``: adds extra options for ``nextpnr-xilinx``.
+		* ``fasm2frames_opts``: adds extra options for ``fasm2frames``.
+		* ``xc7frames2bit_opts``: adds extra options for ``xc7frames2bit``.
```
Comment (lethalbit) [inline torii/platform/vendor/xilinx.py:135]: Please add a new line here so the end of comment has a blank line between it and the last bits of text

**Review by lethalbit — CHANGES_REQUESTED** [review]
LGTM, just the two minor things,

**Comment by lethalbit** [comment]
OH, the changes are small enough you could (probably should) squash these commits

**Review by lethalbit — APPROVED** [review]

**Review by dragonmux — APPROVED** [review]
LGTM! Thank you lak

**Pull request merged**